### PR TITLE
 Initial SNMP Implementation

### PR DESCRIPTION
**Pull request by sjames-iberis**
Supports Get/GetNext/GetBulk/Walk/BulkWalk.

Using https://github.com/geoffgarside/ber for ber encoding/decoding, supplemented by
custom code to intercept decoding of SNMP-specific datatypes.

Still need to:

consider options for 'sugar' to facilitate conversion of SNMP variables
implement API checks to validate OIDs on entry
extend trace hooks, as necessary
implement Close

Current limitations:

SNMPV3 unsupported
tcp network unsupported

### CODE REVIEW

File: v2/netconf/ops/sessionfactory.go [hunk 1]
```diff
@@ -10,14 +10,14 @@ import (
 
 // Defines a factory method for instantiating netconf sessions.
 
-// NewRPCSession connects to the  target using the ssh configuration, and establishes
+// NewSession connects to the  target using the ssh configuration, and establishes
```
Comment (damianoneill) [inline v2/netconf/ops/sessionfactory.go:13]: Assume this break all existing users of the lib?

Comment (sjames-iberis) [reply v2/netconf/ops/sessionfactory.go:13]: No. Has no impact on netconf users.

File: v2/snmp/session.go [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+	// Get Bulk request processing is described at https://tools.ietf.org/html/rfc1905#section-4.2.3
+	GetBulk(ctx context.Context, oids []string, nonRepeaters int, maxRepetitions int) (*PDU, error)
+
+	// Issues SNMP GET NEXT requests starting from the specified root oid, invoking the function walker for each
+	// variable that is a descendant of the root oid.
+	Walk(ctx context.Context, rootOid string, walker Walker) error
+
+	// Issues SNMP GET BULK requests starting from the specified root oid, invoking the function walker for each
+	// variable that is a descendant of the root oid.
+	BulkWalk(ctx context.Context, rootOid string, maxRepetitions int, walker Walker) error
```
Comment (damianoneill) [inline v2/snmp/session.go:34]: param order, above is ctx, oid, walker, should this be ctx, void, walker maxreps

File: v2/netconf/ops/sessionfactory.go (hunk 1 above)
Comment (sjames-iberis) [reply v2/netconf/ops/sessionfactory.go:13]: Sorry - it's only the comment that's been updated.

File: v2/snmp/session.go (hunk 2 above)
Comment (sjames-iberis) [reply v2/snmp/session.go:34]: I took the view that maxReps was more closely related to the oid.

File: v2/snmp/session.go [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+	return m.executeWalk(ctx, getBulkMessage, maxRepetitions, rootOid, walker)
+}
+
+// Generic Get execution.
+// Generates a packet to define the type of Get, the required oids and, in the case of a bulk get, the associated
+// non-repeaters and max-repetitions values.
+// Returns a PDU with the resolved variable bindings.
+func (m *sessionImpl) executeGet(ctx context.Context, getType messageType, oids []string, nonRepeaters, maxRepetitions int) (*PDU, error) {
+
+	// TODO Validate OIDs on entry.
```
Comment (damianoneill) [inline v2/snmp/session.go:129]: are you wanting to leave these todos in?

File: v2/snmp/sessionfactory.go [hunk 4]
```diff
@@ -94,20 +96,20 @@ func LoggingHooks(trace *ManagerTrace) ManagerOption {
 type SNMPVersion int
 
 const (
-	SNMPV1 SNMPVersion = iota
-	SNMPV2C
-	SNMPV3
+	SNMPV1  SNMPVersion = 0
+	SNMPV2C SNMPVersion = 1
```
Comment (damianoneill) [inline v2/snmp/sessionfactory.go:100]: i think this should be 2, most people when talking about 2 mean 2c

File: v2/snmp/types.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+	case Time:
+		return uint32(v)
+
+	case Counter64:
+		return uint64(v)
+	}
+	return v
+}
+
+// Unmarshals an SNMP octetstring-based variable into a TypedValue.
```
Comment (damianoneill) [inline v2/snmp/types.go:131]: should the comment be Unmarshals an String octetstring-based or the function name changed to unmarshalSNMP

File: v2/snmp/sessionfactory.go (hunk 4 above)
Comment (sjames-iberis) [reply v2/snmp/sessionfactory.go:100]: I think that's a protocol issue - has to be 1.

File: v2/snmp/session.go (hunk 3 above)
Comment (sjames-iberis) [reply v2/snmp/session.go:129]: Prefer to, for the time being.

File: v2/snmp/types.go (hunk 5 above)
Comment (sjames-iberis) [reply v2/snmp/types.go:131]: Will rename.  
Also taking opportunity to add a couple of methods to deliver abstract TypedValue as string, int etc.

**Pull request merged**